File: molecular/src/main/java/com/hartwig/actin/molecular/orange/interpretation/CharacteristicsExtraction.java
package com.hartwig.actin.molecular.orange.interpretation;

import java.util.Set;

import com.hartwig.actin.molecular.datamodel.characteristics.ImmutableMolecularCharacteristics;
import com.hartwig.actin.molecular.datamodel.characteristics.ImmutablePredictedTumorOrigin;
import com.hartwig.actin.molecular.datamodel.characteristics.MolecularCharacteristics;
import com.hartwig.actin.molecular.datamodel.characteristics.PredictedTumorOrigin;
import com.hartwig.actin.molecular.orange.datamodel.OrangeRecord;
import com.hartwig.actin.molecular.orange.datamodel.cuppa.CuppaPrediction;
import com.hartwig.actin.molecular.orange.datamodel.purple.PurpleRecord;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

final class CharacteristicsExtraction {

    private static final Logger LOGGER = LogManager.getLogger(CharacteristicsExtraction.class);

    static final String MICROSATELLITE_STABLE = "MSS";
    static final String MICROSATELLITE_UNSTABLE = "MSI";

    static final String HOMOLOGOUS_REPAIR_DEFICIENT = "HR_DEFICIENT";
    static final String HOMOLOGOUS_REPAIR_PROFICIENT = "HR_PROFICIENT";
    static final String HOMOLOGOUS_REPAIR_UNKNOWN = "CANNOT_BE_DETERMINED";

    static final String TUMOR_STATUS_HIGH = "HIGH";
    static final String TUMOR_STATUS_LOW = "LOW";
    static final String TUMOR_STATUS_UNKNOWN = "UNKNOWN";

    static final double HIGH_TMB_CUTOFF = 10;

    private CharacteristicsExtraction() {
    }

    @NotNull
    public static MolecularCharacteristics extract(@NotNull OrangeRecord record) {
        CuppaPrediction best = findBestCuppaPrediction(record.cuppa().predictions());
        PredictedTumorOrigin predictedTumorOrigin = best != null
                ? ImmutablePredictedTumorOrigin.builder().tumorType(best.cancerType()).likelihood(best.likelihood()).build()
                : null;

        PurpleRecord purple = record.purple();

        // TODO: Make TMB interpretation inside ORANGE.

        return ImmutableMolecularCharacteristics.builder()
                .purity(purple.purity())
                .ploidy(purple.ploidy())
                .predictedTumorOrigin(predictedTumorOrigin)
                .isMicrosatelliteUnstable(isMSI(purple.microsatelliteStabilityStatus()))
                .isHomologousRepairDeficient(isHRD(record.chord().hrStatus()))
                .tumorMutationalBurden(purple.tumorMutationalBurden())
                .hasHighTumorMutationalBurden(purple.tumorMutationalBurden() >= HIGH_TMB_CUTOFF)
                .tumorMutationalLoad(purple.tumorMutationalLoad())
                .hasHighTumorMutationalLoad(hasHighStatus(purple.tumorMutationalLoadStatus()))
                .build();
    }

    @Nullable
    private static CuppaPrediction findBestCuppaPrediction(@NotNull Set<CuppaPrediction> predictions) {
        CuppaPrediction best = null;
        for (CuppaPrediction prediction : predictions) {
            if (best == null || prediction.likelihood() > best.likelihood()) {
                best = prediction;
            }
        }
        return best;
    }

    @Nullable
    private static Boolean isMSI(@NotNull String microsatelliteStatus) {
        if (microsatelliteStatus.equals(MICROSATELLITE_UNSTABLE)) {
            return true;
        } else if (microsatelliteStatus.equals(MICROSATELLITE_STABLE)) {
            return false;
        }

        LOGGER.warn("Cannot interpret microsatellite status '{}'", microsatelliteStatus);
        return null;
    }

    @Nullable
    private static Boolean isHRD(@NotNull String hrStatus) {
        switch (hrStatus) {
            case HOMOLOGOUS_REPAIR_DEFICIENT:
                return true;
            case HOMOLOGOUS_REPAIR_PROFICIENT:
                return false;
            case HOMOLOGOUS_REPAIR_UNKNOWN:
                return null;
        }

        LOGGER.warn("Cannot interpret homologous repair status '{}'", hrStatus);
        return null;
    }

    @Nullable
    private static Boolean hasHighStatus(@NotNull String tumorMutationalStatus) {
        switch (tumorMutationalStatus) {
            case TUMOR_STATUS_HIGH: {
                return true;
            }
            case TUMOR_STATUS_LOW: {
                return false;
            }
            case TUMOR_STATUS_UNKNOWN: {
                return null;
            }
        }

        LOGGER.warn("Cannot interpret tumor mutational status: {}", tumorMutationalStatus);
        return null;
    }
}

File: molecular/src/main/kotlin/com/hartwig/actin/molecular/orange/CharacteristicsExtraction.kt
package com.hartwig.actin.molecular.orange

import com.hartwig.actin.datamodel.molecular.characteristics.CupPrediction
import com.hartwig.actin.datamodel.molecular.characteristics.CuppaMode
import com.hartwig.actin.datamodel.molecular.characteristics.HomologousRecombination
import com.hartwig.actin.datamodel.molecular.characteristics.HomologousRecombinationType
import com.hartwig.actin.datamodel.molecular.characteristics.MicrosatelliteStability
import com.hartwig.actin.datamodel.molecular.characteristics.MolecularCharacteristics
import com.hartwig.actin.datamodel.molecular.characteristics.PredictedTumorOrigin
import com.hartwig.actin.datamodel.molecular.characteristics.TumorMutationalBurden
import com.hartwig.actin.datamodel.molecular.characteristics.TumorMutationalLoad
import com.hartwig.actin.molecular.util.ExtractionUtil
import com.hartwig.hmftools.datamodel.chord.ChordRecord
import com.hartwig.hmftools.datamodel.chord.ChordStatus
import com.hartwig.hmftools.datamodel.cuppa.CuppaData
import com.hartwig.hmftools.datamodel.cuppa.CuppaPrediction
import com.hartwig.hmftools.datamodel.orange.OrangeRecord
import com.hartwig.hmftools.datamodel.purple.PurpleCharacteristics
import com.hartwig.hmftools.datamodel.purple.PurpleMicrosatelliteStatus
import com.hartwig.hmftools.datamodel.purple.PurpleTumorMutationalStatus

object CharacteristicsExtraction {

    fun extract(record: OrangeRecord): MolecularCharacteristics {
        val purple = record.purple()

        return MolecularCharacteristics(
            purity = purple.fit().purity(),
            ploidy = purple.fit().ploidy(),
            predictedTumorOrigin = determinePredictedTumorOrigin(record.cuppa()),
            microsatelliteStability = determineMicrosatelliteStability(purple.characteristics()),
            homologousRecombination = determineHomologousRecombination(record.chord()),
            tumorMutationalBurden = determineTumorMutationalBurden(purple.characteristics()),
            tumorMutationalLoad = determineTumorMutationalLoad(purple.characteristics())
        )
    }

    private fun determinePredictedTumorOrigin(cuppa: CuppaData?): PredictedTumorOrigin? {
        return cuppa?.let {
            PredictedTumorOrigin(predictions = determineCupPredictions(it.predictions(), CuppaMode.valueOf(it.mode().toString())))
        }
    }

    private fun determineMicrosatelliteStability(characteristics: PurpleCharacteristics): MicrosatelliteStability? {
        return isMicrosatelliteUnstable(characteristics.microsatelliteStatus())?.let { isUnstable ->
            MicrosatelliteStability(
                microsatelliteIndelsPerMb = characteristics.microsatelliteIndelsPerMb(),
                isUnstable = isUnstable,
                evidence = ExtractionUtil.noEvidence()
            )
        }
    }

    private fun determineHomologousRecombination(chord: ChordRecord?): HomologousRecombination? {
        return chord?.let {
            HomologousRecombination(
                score = it.hrdValue(),
                isDeficient = isHomologousRecombinationDeficient(it.hrStatus()),
                type = HomologousRecombinationType.valueOf(it.hrdType().uppercase()),
                brca1Value = it.brca1Value(),
                brca2Value = it.brca2Value(),
                evidence = ExtractionUtil.noEvidence()
            )
        }
    }

    private fun determineTumorMutationalBurden(characteristics: PurpleCharacteristics): TumorMutationalBurden? {
        return hasHighStatus(characteristics.tumorMutationalBurdenStatus())?.let { isHigh ->
            TumorMutationalBurden(
                score = characteristics.tumorMutationalBurdenPerMb(),
                isHigh = isHigh,
                evidence = ExtractionUtil.noEvidence()
            )
        }
    }

    private fun determineTumorMutationalLoad(characteristics: PurpleCharacteristics): TumorMutationalLoad? {
        return hasHighStatus(characteristics.tumorMutationalLoadStatus())?.let { isHigh ->
            TumorMutationalLoad(
                score = characteristics.tumorMutationalLoad(),
                isHigh = isHigh,
                evidence = ExtractionUtil.noEvidence()
            )
        }
    }

    private fun isMicrosatelliteUnstable(microsatelliteStatus: PurpleMicrosatelliteStatus): Boolean? {
        return when (microsatelliteStatus) {
            PurpleMicrosatelliteStatus.MSI -> true
            PurpleMicrosatelliteStatus.MSS -> false
            PurpleMicrosatelliteStatus.UNKNOWN -> null
        }
    }

    private fun isHomologousRecombinationDeficient(hrStatus: ChordStatus): Boolean? {
        return when (hrStatus) {
            ChordStatus.HR_DEFICIENT -> true
            ChordStatus.HR_PROFICIENT -> false
            ChordStatus.UNKNOWN, ChordStatus.CANNOT_BE_DETERMINED -> null
        }
    }

    private fun hasHighStatus(tumorMutationalStatus: PurpleTumorMutationalStatus): Boolean? {
        return when (tumorMutationalStatus) {
            PurpleTumorMutationalStatus.HIGH -> {
                true
            }

            PurpleTumorMutationalStatus.LOW -> {
                false
            }

            PurpleTumorMutationalStatus.UNKNOWN -> {
                null
            }
        }
    }

    private fun determineCupPredictions(cuppaPredictions: List<CuppaPrediction>, cuppaMode: CuppaMode): List<CupPrediction> {
        return cuppaPredictions.map { cuppaPrediction: CuppaPrediction -> determineCupPrediction(cuppaPrediction, cuppaMode) }
    }

    private fun determineCupPrediction(cuppaPrediction: CuppaPrediction, cuppaMode: CuppaMode): CupPrediction {
        if (cuppaPrediction.snvPairwiseClassifier() == null || cuppaPrediction.genomicPositionClassifier() == null ||
            cuppaPrediction.featureClassifier() == null
        ) {
            throw IllegalStateException(
                "CUPPA classifiers are not supposed to be missing at this point " +
                        "in cuppa prediction: $cuppaPrediction"
            )
        }

        return CupPrediction(
            cancerType = cuppaPrediction.cancerType(),
            likelihood = cuppaPrediction.likelihood(),
            snvPairwiseClassifier = cuppaPrediction.snvPairwiseClassifier()!!,
            genomicPositionClassifier = cuppaPrediction.genomicPositionClassifier()!!,
            featureClassifier = cuppaPrediction.featureClassifier()!!,
            cuppaMode = cuppaMode
        )
    }
}